### PR TITLE
Add hal::delay function, implement delay/wait for esp32

### PR DESCRIPTION
**Pull request by jajik**
This PR

 * adds a new non-blocking variant for `wait` function – `delay`;
 * implements `wait` and `delay` for esp32;
 * fixes non-blocking `wait` behaviour in the SimpleSim (You can check the behaviour using tutorial 2 by adding `wait`/`delay` before the Hello, World!).

### CODE REVIEW

File: softwareComponents/rofiHalInc/include/rofi_hal.hpp [hunk 1]
```diff
@@ -622,14 +622,23 @@ class RoFI {
...
     static void wait( int ms, std::function< void() > callback );
 
+    /**
+     * \brief Call callback after given delay. The call is non-blocking.
+     *
+     * You can assume that the callback will always be called exactly once.
+     * \param ms delay in milliseconds
+     * \param callback non-empty callback to be called after the delay
+     */
+    static void delay( int ms, std::function< void() > callback );
```
Comment (yaqwsx) [inline softwareComponents/rofiHalInc/include/rofi_hal.hpp:640]: The naming and even the contract is very confusing. In my opinion, we should have:
- `void delay( int ms )` that takes no function and simply blocks the caller for a given period of time.
- `void postpone( int ms, std::function< void() > callback )` that returns immediately and calls `callback` after given period of time in unspecified context.

@jajik: What do you think?

Comment (yaqwsx) [inline softwareComponents/rofiHalInc/include/rofi_hal.hpp:640]: The naming and even the contract is very confusing. In my opinion, we should have:
- `void delay( int ms )` that takes no function and simply blocks the caller for a given period of time.
- `void postpone( int ms, std::function< void() > callback )` that returns immediately and calls `callback` after given period of time in unspecified context.

@jajik: What do you think?

Comment (jajik) [reply softwareComponents/rofiHalInc/include/rofi_hal.hpp:640]: If we don't mind renaming `wait`, I'm ok with it. I would personally prefer `delay` -> `sleep` (`wait` would be even better but we should get rid of it if we make these changes) & `postpone` -> `defer`. But that's just a detail.

I'll change it, however that will require a few additional commits fixing whatever will break. I'll try to have it done by tomorrow afternoon.

Comment (yaqwsx) [reply softwareComponents/rofiHalInc/include/rofi_hal.hpp:640]: `delay` or `sleep` - I don't have a preference. You can choose as you want.

**Comment by jajik** [comment]
I changed the functions as proposed. I left it in a separate commit, it seemed more appropriate to me.

**Review by yaqwsx — APPROVED** [review]

**Comment by jajik** [comment]
Merged, thanks!

**Pull request merged**